task_processing/plugins/mesos/execution_framework.py

```diff
@@ -527,20 +568,39 @@ def resourceOffers(self, driver, offers):
 
             accepted.append('offer: {} agent: {} tasks: {}'.format(
                 offer.id.value, offer.agent_id.value, len(tasks_to_launch)))
-            driver.launchTasks(offer.id, tasks_to_launch)
 
+            task_launch_failed = False
+            try:
+                driver.launchTasks(offer.id, tasks_to_launch)
+            except (socket.timeout, Exception):
+                log.warning('Failed to launch following tasks {tasks}.'
+                            'Thus, moving them to UNKNOWN state'.format(
+                                tasks=', '.join([
+                                    task.task_id.value for task in
+                                    tasks_to_launch
+                                ]),
+                            )
+                            )
+                task_launch_failed = True
+
+            # 'UNKNOWN' state is for internal tracking. It will not be
+            # propogated to users.
+            current_task_state = 'UNKNOWN' if task_launch_failed else \
+                'TASK_STAGING'
             with self._lock:
                 for task in tasks_to_launch:
                     md = self.task_metadata[task.task_id.value]
                     self.task_metadata = self.task_metadata.set(
                         task.task_id.value,
                         md.set(
-                            task_state='TASK_STAGING',
+                            task_state=current_task_state,
                             task_state_history=md.task_state_history.set(
-                                'TASK_STAGING', time.time()),
+                                current_task_state, time.time()),
+
                         )
                     )
-                    get_metric(TASK_LAUNCHED_COUNT).count(1)
+                    if not task_launch_failed:
```

Comment:
we should probably be counting number of failures as well, sounds like very useful statistic

Reply:
Good call. Added a metric for this.